Improve the upgrading documentation

ptgott opened the pull request:
Closes #33817

This change restructures the information architecture of learning how to upgrade Teleport. Currently, users who want to upgrade Teleport need to cobble together instructions from pages in various subsections of the docs site.

With this change, users can navigate to a single subsection of the "Home" section, "Upgrading", and follow a linear flow for their Teleport edition (self-hosted or Cloud) and platform (Kubernetes/Linux servers).

The new information architecture replaces the following guides:

- docs/pages/management/admin/upgrading-the-teleport-binary.mdx
- docs/pages/management/operations/enroll-agent-into-automatic-updates.mdx
- docs/pages/management/operations/upgrading.mdx

The compatibility and upgrade sequence information in `upgrading.mdx` is now in a separate page that readers are expected to be familiar with before completing the how-to guides in this subsection.

As part of the change, the guide for setting up self-hosted automatic agent upgrade infrastructure is now in the "Upgrading" subsection. This guide now includes example Terraform configs for deploying a version server. Since there are many ways to deploy a static file server, the intention is to remove decision making to get the user more quickly to the objective of the guide.

This change also adds an `install-linux-self-hosted.mdx` partial for guides that only need installation instructions for OSS and Enterprise.

Issues closed by these changes:

- Edit Linux server instructions to include commands for upgrading Teleport via a package manager, which closes #32553.
- Removes the separation between the "cluster" and "binary" upgrading docs (closes #31446).
- Organizes the docs by deployment, assuming the user is upgrading an entire cluster at a time (closes #27672, closes #10917).
- By including package manager instructions already present in partials, this change makes it clearer how to update a package manager repo to change major versions (closes #26499).
- Removes the instructions to use SIGUSR2 when upgrading (closes #33720).

ptgott committed:
Improve the upgrading documentation

Closes #33817

This change restructures the information architecture of learning how to
upgrade Teleport. Currently, users who want to upgrade Teleport need to
cobble together instructions from pages in various subsections of the
docs site.

With this change, users can navigate to a single subsection of the
"Home" section, "Upgrading", and follow a linear flow for their Teleport
edition (self-hosted or Cloud) and platform (Kubernetes/Linux servers).

The new information architecture replaces the following guides:

- docs/pages/management/admin/upgrading-the-teleport-binary.mdx
- docs/pages/management/operations/enroll-agent-into-automatic-updates.mdx
- docs/pages/management/operations/upgrading.mdx

The compatibility and upgrade sequence information in `upgrading.mdx` is
now in a separate page that readers are expected to be familiar with
before completing the how-to guides in this subsection.

As part of the change, the guide for setting up self-hosted automatic
agent upgrade infrastructure is now in the "Upgrading" subsection. This
guide now includes example Terraform configs for deploying a version
server. Since there are many ways to deploy a static file server, the
intention is to remove decision making to get the user more quickly to
the objective of the guide.

This change also adds an `install-linux-self-hosted.mdx` partial for
guides that only need installation instructions for OSS and Enterprise.

Issues closed by these changes:

- Edit Linux server instructions to include commands for upgrading
  Teleport via a package manager, which closes #32553.
- Removes the separation between the "cluster" and "binary" upgrading
  docs (closes #31446).
- Organizes the docs by deployment, assuming the user is upgrading an
  entire cluster at a time (closes #27672, closes #10917).
- By including package manager instructions already present in partials,
  this change makes it clearer how to update a package manager repo to
  change major versions (closes #26499).
- Removes the instructions to use SIGUSR2 when upgrading (closes
  #33720).

espadolini commented on docs/pages/upgrading/cloud-kubernetes.mdx:
````diff
@@ -0,0 +1,135 @@
+
+## Step 1/2. Determine whether any agents require automatic upgrades
+
+To determine if any agents in your cluster are not configured for automatic
+upgrades, run the following command. This prints a list of all Teleport services
+with no upgrader, filtering these to ignore the Auth Service and Proxy Service
+(which are maintained for you in Teleport Cloud):
+
+```code
+$ tctl inventory ls --upgrader=none | awk '!($3 ~ /(Proxy|Auth)/){print $0}'
+Server ID                            Hostname                        Services Version Upgrader
+------------------------------------ ------------------------------- -------- ------- --------
+00000000-0000-0000-0000-000000000000 teleport-agent-0                Kube     v13.4.3 none
+```
````
@fspmarshall could we just lie and report a `managed` or `cloud` upgrader for cloud auths and proxies? or support negation in the `--services` filter so we can just exclude auth and proxy agents from the listing without resorting to awk?

espadolini commented on docs/pages/upgrading/cloud-kubernetes.mdx:
```diff
@@ -0,0 +1,135 @@
+
+## Step 2/2. Enroll agents in automatic upgrades
+
+This section assumes that the name of your `teleport-kube-agent` release is
+`teleport-agent`, and that you have installed it in the `teleport-agent`
+namespace.
```
Is this preferrable to using `<Var>` to allow for slightly easier copypasting?

espadolini commented on docs/pages/upgrading/cloud-linux.mdx:
```diff
@@ -0,0 +1,179 @@
+---
+title: Upgrade Teleport Cloud Clusters on Linux
+description: Provides instructions for upgrading Teleport Cloud clusters that run on Linux servers. 
```
same question as the header for cloud-kubernetes.mdx

espadolini commented on docs/pages/upgrading/cloud-linux.mdx:
````diff
@@ -0,0 +1,179 @@
+agent services:
+
+```code
+$ tctl inventory ls \
+  --services=node,kube,db,app,windows_desktop,discovery,okta \
+  --upgrader=none
````
Why the inconsistency between this and the teleport-kube-agent instructions where we filter out the text output with `awk` instead?

ptgott replied:
I think I used this approach after forgetting I'd used `awk` in the other page. I'll use this one in both pages—it's not as concise as a negated service filter, but at least it doesn't require an external tool

espadolini commented on docs/pages/upgrading/cloud-linux.mdx:
````diff
@@ -0,0 +1,179 @@
+```code
+$ HOST=00000000-0000-0000-0000-000000000000
+$ USER=root
+$ tsh ssh ${USER?}@$(tctl get node/${HOST?} --format json | jq -r '.[0].spec.hostname')
````
You can just `tsh ssh <user>@<uuid>` without having to figure out a hostname for the node.
```suggestion
$ tsh ssh "${USER?}@${HOST?}"
```

espadolini commented on docs/pages/upgrading/overview.mdx:
```diff
@@ -0,0 +1,66 @@
+
+## Upgrading multiple Teleport clusters
+
+1. Upgrade the root cluster—that is, the cluster that other clusters
+   trust—first.
+1. Verify the upgrade was successful.
+1. Upgrade the trusted leaf clusters.
```
Maybe we should mention that when upgrading across multiple major versions this must be done one major version at a time - i.e. it's not legal to upgrade the root from v5 to v14 and then the leaf from v5 to v14.

espadolini commented on docs/pages/upgrading/overview.mdx:
```diff
@@ -0,0 +1,66 @@
+1. Upgrade your Teleport agents to the same version number as the Auth Service.
+   You can upgrade resource agents in any sequence or at the same time.
+   
+   If you are upgrading more then one version number, repeat these steps until
+   you reach your target major version number.
```
Should we go ahead and recommend a full cluster rebuild for a cluster that's more than 4 or so major versions behind?

ptgott replied:
What are the steps involved in a full cluster rebuild? Back up dynamic resources, then redeploy the Auth/Proxy and agents?

espadolini replied:
Yes, optionally configuring agents to use a different data directory so you can use the old cluster to access nodes to set up the new cluster (if there's no other external access).

espadolini commented on docs/pages/upgrading/self-hosted-kubernetes.mdx:
```diff
@@ -0,0 +1,153 @@
+1. Update the Teleport Helm chart repository so you can install the latest
+   version of the `teleport-cluster` chart:
+
+    (!docs/pages//kubernetes-access/helm/includes/helm-repo-add.mdx!)
```
```suggestion
    (!docs/pages/kubernetes-access/helm/includes/helm-repo-add.mdx!)
```

espadolini reviewed (commented):
I don't have experience with the automatic upgrader, so I can't really vouch for that part.

zmb3 commented:
@fspmarshall @hugoShaka would you mind reviewing from an auto-upgrades perspective?

hugoShaka commented on docs/pages/upgrading/self-hosted-kubernetes.mdx:
```diff
@@ -0,0 +1,153 @@
+of the Auth Service.
+
+
+## Step 3/3. Enroll agents in automatic upgrades
```
I don't think setting up automatic updates should be the next step in self-hosted setups because:
- we don't support automatic updates for Teleport OSS (because of some sad packaging issue, there's no technical or business limitation)
- most self-hosted users don't have automatic updates version servers. Teleport is not yet hosting its own version server (I hope it will become one day, so we can use automatic updates everywhere).
- there are still a lot of setups not supported by automatic upgrades

hugoShaka commented on docs/pages/upgrading/self-hosted-kubernetes.mdx:
````diff
@@ -0,0 +1,153 @@
+1. Upgrade the Helm release:
+
+    ```code
+    $ helm upgrade teleport-cluster teleport/teleport-cluster --values=values.yaml
````
Regarding the helm upgrade commands (comments valid for all `helm` upgrade guides):
- `helm repo update` should be run before (because if we don't specify it we'll definitely have support tickets about folks not seeing the new release)
- Choosing the version with the `--version` flag.
  ```shell
  helm upgrade <release-name> teleport/teleport-cluster --version <THE VERSION YOU WANT>
  ```
  If we don't explicitly set this we'll end up with 2 issues:
  - users installing the wrong version (surprise downgrade, or 2+ major upgrades)
  - users using the `teleportVersionOverride` Helm chart value to pin the version (we really don't want this to happen, this is a footgun)
- nit: it might be nice to make the release name a variable, because a lot of customers changed it and it's not necessarily clear which part of the command is the release name and should be changed.

hugoShaka commented on docs/pages/upgrading/self-hosted-linux.mdx:
```diff
@@ -0,0 +1,155 @@
+
+   (!docs/pages/includes/start-teleport.mdx!)
+
+## Step 2/2. Enroll agents in automatic upgrades
```
Same comment than for self-hosted kube, I'm not sure automatic upgrades should be the default path for self-hosted.

ptgott commented:
TODO, responding to PR feedback:

- [ ] Respond to straightforward review comments (implementing suggestions or adding detail)
- [ ] Replace Helm release names and Kubernetes namespace names with `Var` components
- [ ] Organize upgrade docs so:
    - Self-hosted users are directed to manual agent upgrade instructions
    - Self-hosted users still have the option of reading version server/automatic agent upgrade instructions
    - Also note that you can't use automatic agent upgrades in OSS Teleport and that there are some caveats.

ptgott committed:
Partially respond to feedback

- Change titles/descriptions (espadolini)
- Avoid using `awk` in `tctl inventory ls` command (espadolini)
- Simplify `tsh ssh` command (espadolini)
- Remove extra slash (espadolini)
- Add a note re: upgrading trusted clusters one major version at a time
  (espadolini)
- Add the `version` flag to the `helm upgrade` command (hugoShaka)

ptgott committed:
Replace release/namespace names with Vars

This responds to PR feedback.

ptgott committed:
Reorganize the upgrading docs

Responds to hugoShaka feedback.

Move all automatic upgrade instructions from the self-hosted upgrading
guides into a single page. The page includes instructions for setting up
automatic upgrade infrastructure as well as enrolling agents on Linux
and Kubernetes. It uses separate H2s for each compute platform, instead
of tabs, to make each section easier to read, since instructions differ
considerably between platforms.
